File: php/src/telegram.php
<?php

declare(strict_types=1);

/** @return array<string,mixed> */
function telegram_parse_event(array $payload): array
{
    $message = $payload['message'] ?? $payload['edited_message'] ?? [];
    $userId = $message['from']['id'] ?? null; $chatId = $message['chat']['id'] ?? null;
    if ($userId === null || $chatId === null) throw new InvalidArgumentException('Telegram update has no supported message');
    $event = ['platform' => 'telegram', 'user_id' => (string)$userId, 'reply_target' => (string)$chatId];
    if (isset($message['text'])) return $event + ['content_type' => 'text', 'text' => $message['text']];
    if (!empty($message['photo'])) { $photo = end($message['photo']); return $event + ['content_type' => 'image', 'media_url' => 'telegram:'.$photo['file_id']]; }
    foreach ([['audio','audio'], ['video','video'], ['document','file']] as [$key, $type]) if (isset($message[$key])) return $event + ['content_type' => $type, 'media_url' => 'telegram:'.$message[$key]['file_id']];
    throw new InvalidArgumentException('Telegram content type is not supported');
}

function telegram_send_reply(array $event, array $reply, string $token): void
{
    if ($token === '') throw new InvalidArgumentException('TELEGRAM_BOT_TOKEN is required');
    foreach (array_slice($reply['messages'] ?? [], 0, 10) as $message) {
        $type = $message['type'] ?? 'text'; $map = ['image' => ['sendPhoto','photo'], 'audio' => ['sendAudio','audio'], 'video' => ['sendVideo','video'], 'file' => ['sendDocument','document']];
        if ($type === 'text') $body = ['chat_id' => $event['reply_target'] ?? $event['user_id'], 'text' => $message['text'] ?? ''];
        else { if (!isset($map[$type], $message['media_url'])) throw new InvalidArgumentException("Telegram {$type} reply requires media_url"); [$method, $field] = $map[$type]; $body = ['chat_id' => $event['reply_target'] ?? $event['user_id'], $field => preg_replace('/^telegram:/', '', $message['media_url'])]; if (!empty($message['text'])) $body['caption'] = $message['text']; }
        $method ??= 'sendMessage'; $url = 'https://api.telegram.org/bot'.$token.'/'.($method ?? 'sendMessage'); $ch = adapter_curl_init($url); curl_setopt_array($ch, [CURLOPT_POST => true, CURLOPT_RETURNTRANSFER => true, CURLOPT_HTTPHEADER => ['Content-Type: application/json'], CURLOPT_POSTFIELDS => json_encode($body, JSON_THROW_ON_ERROR)]); $result = curl_exec($ch); $status = curl_getinfo($ch, CURLINFO_RESPONSE_CODE); if ($result === false || $status < 200 || $status >= 300) throw new RuntimeException('Telegram API request failed'); curl_close($ch);
    }
}

File: go/telegram.go
package common

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
)

type TelegramAdapter struct { Token string; Client *http.Client }

func (a TelegramAdapter) ParseEvent(payload map[string]any) (InboundEvent, error) {
	message, _ := payload["message"].(map[string]any); if message == nil { message, _ = payload["edited_message"].(map[string]any) }
	from, _ := message["from"].(map[string]any); chat, _ := message["chat"].(map[string]any); user, _ := from["id"].(float64); chatID, _ := chat["id"].(float64); if user == 0 || chatID == 0 { return InboundEvent{}, fmt.Errorf("Telegram update has no supported message") }
	event := InboundEvent{Platform:"telegram", UserID:fmt.Sprintf("%.0f", user), ReplyTarget:fmt.Sprintf("%.0f", chatID)}
	if text, ok := message["text"].(string); ok { event.ContentType="text"; event.Text=text; return event, nil }
	if photos, ok := message["photo"].([]any); ok && len(photos)>0 { last, _ := photos[len(photos)-1].(map[string]any); event.ContentType="image"; event.MediaURL="telegram:"+last["file_id"].(string); return event,nil }
	for _, item := range []struct{ key, kind string }{{"audio","audio"},{"video","video"},{"document","file"}} { if value, ok := message[item.key].(map[string]any); ok { event.ContentType=item.kind; event.MediaURL="telegram:"+value["file_id"].(string); return event,nil } }
	return InboundEvent{}, fmt.Errorf("Telegram content type is not supported")
}

func (a TelegramAdapter) SendReply(event InboundEvent, reply OutboundReply) error {
	for i, message := range reply.Messages { if i >= 10 { break }; method, field := "sendMessage", "text"; if message.Type != "text" { methods := map[string][2]string{"image":{"sendPhoto","photo"},"audio":{"sendAudio","audio"},"video":{"sendVideo","video"},"file":{"sendDocument","document"}}; pair, ok := methods[message.Type]; if !ok || message.MediaURL=="" { return fmt.Errorf("unsupported Telegram reply") }; method,field=pair[0],pair[1] }; value:=message.Text; if message.Type!="text" { value=strings.TrimPrefix(message.MediaURL, "telegram:") }; body:=map[string]any{"chat_id":event.ReplyTarget}; body[field]=value; if message.Type!="text"&&message.Text!="" { body["caption"]=message.Text }; data,_:=json.Marshal(body); req,_:=http.NewRequest(http.MethodPost,"https://api.telegram.org/bot"+a.Token+"/"+method,strings.NewReader(string(data))); req.Header.Set("Content-Type","application/json"); client:=a.Client;if client==nil{client=http.DefaultClient}; resp,err:=client.Do(req);if err!=nil{return err};resp.Body.Close();if resp.StatusCode/100!=2{return fmt.Errorf("Telegram API returned %s",resp.Status)} }
	return nil
}
